In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/zomato-bangalore-restaurants/zomato.csv


In [2]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [3]:
df = pd.read_csv("/kaggle/input/zomato-bangalore-restaurants/zomato.csv")
df.head(10)

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   
5  https://www.zomato.com/bangalore/timepass-dinn...   
6  https://www.zomato.com/bangalore/rosewood-inte...   
7  https://www.zomato.com/bangalore/onesta-banash...   
8  https://www.zomato.com/bangalore/penthouse-caf...   
9  https://www.zomato.com/bangalore/smacznego-ban...   

                                             address  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...   
2  1112, Next to KIMS Medical College, 17th Cross...   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...   
5  37, 5-1, 4th Floor, Bosco Court, Gandhi Bazaar...   
6  19/1, New Timberyard Layout, Beside Satellite ...   
7  2469, 3rd Floor, 24th Cross, Opposite BDA Comp...   
8  1, 30th Main Road, 3rd Stage, Banashankari, Ba...   
9  2470, 21 Main Road, 25th Cross, Banashankari, ...   

                                              name online_order book_table  \
0                                            Jalsa          Yes        Yes   
1                                   Spice Elephant          Yes         No   
2                                  San Churro Cafe          Yes         No   
3                            Addhuri Udupi Bhojana           No         No   
4                                    Grand Village           No         No   
5                                  Timepass Dinner          Yes         No   
6  Rosewood International Hotel - Bar & Restaurant           No         No   
7                                           Onesta          Yes        Yes   
8                                   Penthouse Cafe          Yes         No   
9                                        Smacznego          Yes         No   

    rate  votes                             phone      location  \
0  4.1/5    775    080 42297555\r\n+91 9743772233  Banashankari   
1  4.1/5    787                      080 41714161  Banashankari   
2  3.8/5    918                    +91 9663487993  Banashankari   
3  3.7/5     88                    +91 9620009302  Banashankari   
4  3.8/5    166  +91 8026612447\r\n+91 9901210005  Basavanagudi   
5  3.8/5    286  +91 9980040002\r\n+91 9980063005  Basavanagudi   
6  3.6/5      8    +91 9731716688\r\n080 26740366   Mysore Road   
7  4.6/5   2556      080 48653961\r\n080 48655715  Banashankari   
8  4.0/5    324  +91 8884135549\r\n+91 9449449316  Banashankari   
9  4.2/5    504  +91 9945230807\r\n+91 9743804471  Banashankari   

             rest_type                                         dish_liked  \
0        Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1        Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3          Quick Bites                                        Masala Dosa   
4        Casual Dining                                Panipuri, Gol Gappe   
5        Casual Dining  Onion Rings, Pasta, Kadhai Paneer, Salads, Sal...   
6        Casual Dining                                                NaN   
7  Casual Dining, Cafe  Farmhouse Pizza, Chocolate Banana, Virgin Moji...   
8                 Cafe  Pizza, Mocktails, Coffee, Nachos, Salad, Pasta...   
9                 Cafe  Waffles, Pasta, Coleslaw Sandwich, Choco Waffl...   

                                      cuisines approx_cost(for two people)  \
0               North Indian, Mughlai, Chinese                         800   
1                  Chinese, North Indian, Thai                         800   
2                       Cafe, Mexican, Italian                         800   
3                   South Indian, North Indian            

In [4]:
df.shape 
# No. of rows = 51717 and No. of Columns = 17

(51717, 17)

In [5]:
df.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

In [6]:
# Removing unwanted columns
df = df.drop(['url','address','phone','menu_item', 'dish_liked','reviews_list'], axis =1)

In [7]:
df.head(10)

name online_order book_table  \
0                                            Jalsa          Yes        Yes   
1                                   Spice Elephant          Yes         No   
2                                  San Churro Cafe          Yes         No   
3                            Addhuri Udupi Bhojana           No         No   
4                                    Grand Village           No         No   
5                                  Timepass Dinner          Yes         No   
6  Rosewood International Hotel - Bar & Restaurant           No         No   
7                                           Onesta          Yes        Yes   
8                                   Penthouse Cafe          Yes         No   
9                                        Smacznego          Yes         No   

    rate  votes      location            rest_type  \
0  4.1/5    775  Banashankari        Casual Dining   
1  4.1/5    787  Banashankari        Casual Dining   
2  3.8/5    918  Banashankari  Cafe, Casual Dining   
3  3.7/5     88  Banashankari          Quick Bites   
4  3.8/5    166  Basavanagudi        Casual Dining   
5  3.8/5    286  Basavanagudi        Casual Dining   
6  3.6/5      8   Mysore Road        Casual Dining   
7  4.6/5   2556  Banashankari  Casual Dining, Cafe   
8  4.0/5    324  Banashankari                 Cafe   
9  4.2/5    504  Banashankari                 Cafe   

                                      cuisines approx_cost(for two people)  \
0               North Indian, Mughlai, Chinese                         800   
1                  Chinese, North Indian, Thai                         800   
2                       Cafe, Mexican, Italian                         800   
3                   South Indian, North Indian                         300   
4                     North Indian, Rajasthani                         600   
5                                 North Indian                         600   
6  North Indian, South Indian, Andhra, Chinese                         800   
7                         Pizza, Cafe, Italian                         600   
8                   Cafe, Italian, Continental                         700   
9     Cafe, Mexican, Italian, Momos, Beverages                         550   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari  
5          Buffet    Banashankari  
6          Buffet    Banashankari  
7           Cafes    Banashankari  
8           Cafes    Banashankari  
9           Cafes    Banashankari

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   name                         51717 non-null  object
 1   online_order                 51717 non-null  object
 2   book_table                   51717 non-null  object
 3   rate                         43942 non-null  object
 4   votes                        51717 non-null  int64 
 5   location                     51696 non-null  object
 6   rest_type                    51490 non-null  object
 7   cuisines                     51672 non-null  object
 8   approx_cost(for two people)  51371 non-null  object
 9   listed_in(type)              51717 non-null  object
 10  listed_in(city)              51717 non-null  object
dtypes: int64(1), object(10)
memory usage: 4.3+ MB


Null Values present for columns - 
* rate 
* location
* rest_type
* cuisines
* approx_cost(for two people)

In [9]:
# Removing duplicate rows if any
df.drop_duplicates(inplace = True)
df.shape

(51609, 11)

No. of dupicate rows = (51717 - 51609) = 108 rows removed.

# Cleaning Rate Column

In [10]:
df.rate.unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

# Removing 'NEW', '-', '/5' from rate column

In [11]:
import re

In [12]:

def rate_clean(x):
    if (x=='-' or x=='NEW'):
        return np.nan
        
    else:
        m= re.split('/',str(x))
        return float(m[0])
    
        

df['rate'] = df['rate'].apply(rate_clean)

df.head()
    

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines  \
0        Casual Dining  North Indian, Mughlai, Chinese   
1        Casual Dining     Chinese, North Indian, Thai   
2  Cafe, Casual Dining          Cafe, Mexican, Italian   
3          Quick Bites      South Indian, North Indian   
4        Casual Dining        North Indian, Rajasthani   

  approx_cost(for two people) listed_in(type) listed_in(city)  
0                         800          Buffet    Banashankari  
1                         800          Buffet    Banashankari  
2                         800          Buffet    Banashankari  
3                         300          Buffet    Banashankari  
4                         600          Buffet    Banashankari

# Replace null values in rate column with mean value

In [13]:
df['rate'].fillna(df['rate'].mean(),inplace =True)

In [14]:
# Check if all nulls are replaced
df['rate'].isnull().sum()

0

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51609 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   name                         51609 non-null  object 
 1   online_order                 51609 non-null  object 
 2   book_table                   51609 non-null  object 
 3   rate                         51609 non-null  float64
 4   votes                        51609 non-null  int64  
 5   location                     51588 non-null  object 
 6   rest_type                    51382 non-null  object 
 7   cuisines                     51564 non-null  object 
 8   approx_cost(for two people)  51265 non-null  object 
 9   listed_in(type)              51609 non-null  object 
 10  listed_in(city)              51609 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 4.7+ MB


# Dropping null values from :
* location
* rest_type
* cuisines
* approx_cost(for two people)
 # as information is missing and can't be derived through any statistical assumptions.

In [16]:
df.dropna(inplace=True)
df.info()
# All null values handled.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51042 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   name                         51042 non-null  object 
 1   online_order                 51042 non-null  object 
 2   book_table                   51042 non-null  object 
 3   rate                         51042 non-null  float64
 4   votes                        51042 non-null  int64  
 5   location                     51042 non-null  object 
 6   rest_type                    51042 non-null  object 
 7   cuisines                     51042 non-null  object 
 8   approx_cost(for two people)  51042 non-null  object 
 9   listed_in(type)              51042 non-null  object 
 10  listed_in(city)              51042 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 4.7+ MB


In [17]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines  \
0        Casual Dining  North Indian, Mughlai, Chinese   
1        Casual Dining     Chinese, North Indian, Thai   
2  Cafe, Casual Dining          Cafe, Mexican, Italian   
3          Quick Bites      South Indian, North Indian   
4        Casual Dining        North Indian, Rajasthani   

  approx_cost(for two people) listed_in(type) listed_in(city)  
0                         800          Buffet    Banashankari  
1                         800          Buffet    Banashankari  
2                         800          Buffet    Banashankari  
3                         300          Buffet    Banashankari  
4                         600          Buffet    Banashankari

In [18]:
df.rename(columns={'approx_cost(for two people)':'Cost 2 person', 'listed_in(type)':'type','listed_in(city)':'city'},inplace=True)

In [19]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines Cost 2 person    type  \
0        Casual Dining  North Indian, Mughlai, Chinese           800  Buffet   
1        Casual Dining     Chinese, North Indian, Thai           800  Buffet   
2  Cafe, Casual Dining          Cafe, Mexican, Italian           800  Buffet   
3          Quick Bites      South Indian, North Indian           300  Buffet   
4        Casual Dining        North Indian, Rajasthani           600  Buffet   

           city  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

# Cleaning and converting 'Cost 2 person' column to float datatype

In [20]:
df['Cost 2 person'].unique()

array(['800', '300', '600', '700', '550', '500', '450', '650', '400',
       '900', '200', '750', '150', '850', '100', '1,200', '350', '250',
       '950', '1,000', '1,500', '1,300', '199', '80', '1,100', '160',
       '1,600', '230', '130', '50', '190', '1,700', '1,400', '180',
       '1,350', '2,200', '2,000', '1,800', '1,900', '330', '2,500',
       '2,100', '3,000', '2,800', '3,400', '40', '1,250', '3,500',
       '4,000', '2,400', '2,600', '120', '1,450', '469', '70', '3,200',
       '60', '560', '240', '360', '6,000', '1,050', '2,300', '4,100',
       '5,000', '3,700', '1,650', '2,700', '4,500', '140'], dtype=object)

In [21]:

def conv_cost(i):
    z=str(i)
    z=z.replace(',', '')
    return float(z)

df['Cost 2 person'] = df['Cost 2 person'].apply(conv_cost)

df.info()
    

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51042 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           51042 non-null  object 
 1   online_order   51042 non-null  object 
 2   book_table     51042 non-null  object 
 3   rate           51042 non-null  float64
 4   votes          51042 non-null  int64  
 5   location       51042 non-null  object 
 6   rest_type      51042 non-null  object 
 7   cuisines       51042 non-null  object 
 8   Cost 2 person  51042 non-null  float64
 9   type           51042 non-null  object 
 10  city           51042 non-null  object 
dtypes: float64(2), int64(1), object(8)
memory usage: 4.7+ MB


In [22]:
df['Cost 2 person'].unique

<bound method Series.unique of 0         800.0
1         800.0
2         800.0
3         300.0
4         600.0
          ...  
51712    1500.0
51713     600.0
51714    2000.0
51715    2500.0
51716    1500.0
Name: Cost 2 person, Length: 51042, dtype: float64>

In [23]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

             rest_type                        cuisines  Cost 2 person    type  \
0        Casual Dining  North Indian, Mughlai, Chinese          800.0  Buffet   
1        Casual Dining     Chinese, North Indian, Thai          800.0  Buffet   
2  Cafe, Casual Dining          Cafe, Mexican, Italian          800.0  Buffet   
3          Quick Bites      South Indian, North Indian          300.0  Buffet   
4        Casual Dining        North Indian, Rajasthani          600.0  Buffet   

           city  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

# Clean "rest_type" column

In [24]:
df.rest_type.value_counts()


Quick Bites                   19010
Casual Dining                 10253
Cafe                           3682
Delivery                       2574
Dessert Parlor                 2242
                              ...  
Dessert Parlor, Kiosk             2
Food Court, Beverage Shop         2
Dessert Parlor, Food Court        2
Quick Bites, Kiosk                1
Sweet Shop, Dessert Parlor        1
Name: rest_type, Length: 93, dtype: int64

# Club lesser no. of restaurant categories into "others". Categories having restaurants less than 1000

In [25]:
rest_types_count = df.rest_type.value_counts()
rest_less_1000 = rest_types_count[rest_types_count<1000]
print(rest_less_1000)

Beverage Shop                 863
Bar                           686
Food Court                    616
Sweet Shop                    468
Bar, Casual Dining            411
                             ... 
Dessert Parlor, Kiosk           2
Food Court, Beverage Shop       2
Dessert Parlor, Food Court      2
Quick Bites, Kiosk              1
Sweet Shop, Dessert Parlor      1
Name: rest_type, Length: 85, dtype: int64


In [26]:
rest_less_1000.index

Index(['Beverage Shop', 'Bar', 'Food Court', 'Sweet Shop',
       'Bar, Casual Dining', 'Lounge', 'Pub', 'Fine Dining',
       'Casual Dining, Cafe', 'Beverage Shop, Quick Bites',
       'Bakery, Quick Bites', 'Mess', 'Pub, Casual Dining',
       'Sweet Shop, Quick Bites', 'Kiosk', 'Dessert Parlor, Cafe',
       'Cafe, Casual Dining', 'Cafe, Bakery', 'Cafe, Dessert Parlor',
       'Bakery, Dessert Parlor', 'Microbrewery, Casual Dining',
       'Casual Dining, Pub', 'Dessert Parlor, Quick Bites', 'Takeaway',
       'Cafe, Quick Bites', 'Quick Bites, Beverage Shop',
       'Food Court, Quick Bites', 'Dessert Parlor, Bakery',
       'Quick Bites, Dessert Parlor', 'Food Truck', 'Pub, Microbrewery',
       'Dessert Parlor, Beverage Shop', 'Beverage Shop, Dessert Parlor',
       'Bakery, Cafe', 'Quick Bites, Sweet Shop', 'Quick Bites, Bakery',
       'Casual Dining, Microbrewery', 'Microbrewery, Pub', 'Confectionery',
       'Quick Bites, Cafe', 'Pub, Cafe', 'Dhaba', 'Pub, Bar',
       'Fine

In [27]:
def category(i):
    if i in rest_less_1000.index:
        return "Others"
    else:
        return i
    
df.rest_type = df.rest_type.apply(category)   

df.rest_type.value_counts()
    
        


Quick Bites           19010
Casual Dining         10253
Others                 9003
Cafe                   3682
Delivery               2574
Dessert Parlor         2242
Takeaway, Delivery     2008
Bakery                 1140
Casual Dining, Bar     1130
Name: rest_type, dtype: int64

# Location is good enough for our data analysis. City is redundant column. So, eliminating City column

In [28]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

       rest_type                        cuisines  Cost 2 person    type  \
0  Casual Dining  North Indian, Mughlai, Chinese          800.0  Buffet   
1  Casual Dining     Chinese, North Indian, Thai          800.0  Buffet   
2         Others          Cafe, Mexican, Italian          800.0  Buffet   
3    Quick Bites      South Indian, North Indian          300.0  Buffet   
4  Casual Dining        North Indian, Rajasthani          600.0  Buffet   

           city  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

In [29]:
df.drop("city",axis=1,inplace=True)
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

       rest_type                        cuisines  Cost 2 person    type  
0  Casual Dining  North Indian, Mughlai, Chinese          800.0  Buffet  
1  Casual Dining     Chinese, North Indian, Thai          800.0  Buffet  
2         Others          Cafe, Mexican, Italian          800.0  Buffet  
3    Quick Bites      South Indian, North Indian          300.0  Buffet  
4  Casual Dining        North Indian, Rajasthani          600.0  Buffet

# Clubbing Location having less than 300 restaurants as Others

In [30]:
rest_loc_group = df.groupby('location').size()
rest_loc_group

location
BTM                  5056
Banashankari          902
Banaswadi             640
Bannerghatta Road    1609
Basavanagudi          684
                     ... 
West Bangalore          6
Whitefield           2105
Wilson Garden         246
Yelahanka               5
Yeshwantpur           119
Length: 93, dtype: int64

In [31]:
rest_less_300 = rest_loc_group[rest_loc_group<300]
rest_less_300

location
Basaveshwara Nagar               187
Bommanahalli                     236
CV Raman Nagar                    89
Central Bangalore                  8
City Market                      122
East Bangalore                    43
HBR Layout                       153
Hebbal                            14
Hennur                           159
Hosur Road                        98
ITPL Main Road, Whitefield       113
Infantry Road                    150
Jakkur                             3
Jalahalli                         38
Jeevan Bhima Nagar               268
KR Puram                          18
Kaggadasapura                    101
Kanakapura Road                   19
Kengeri                            8
Koramangala                       48
Koramangala 2nd Block            102
Koramangala 3rd Block            215
Koramangala 8th Block            294
Kumaraswamy Layout               191
Langford Town                     27
Magadi Road                       34
Majestic                     

In [32]:
def loc_category(i):
    if i in rest_less_300.index:
        return "Others"
    else:
        return i
    
df.location = df.location.apply(loc_category)   

df.groupby('location').size()

location
BTM                      5056
Banashankari              902
Banaswadi                 640
Bannerghatta Road        1609
Basavanagudi              684
Bellandur                1268
Brigade Road             1210
Brookefield               656
Church Street             566
Commercial Street         370
Cunningham Road           490
Domlur                    482
Ejipura                   433
Electronic City          1246
Frazer Town               714
HSR                      2494
Indiranagar              2026
JP Nagar                 2218
Jayanagar                1916
Kalyan Nagar              841
Kammanahalli              639
Koramangala 1st Block    1236
Koramangala 4th Block    1017
Koramangala 5th Block    2479
Koramangala 6th Block    1127
Koramangala 7th Block    1174
Lavelle Road              518
MG Road                   893
Malleshwaram              721
Marathahalli             1805
New BEL Road              644
Old Airport Road          437
Others                   4954
R

In [33]:
rest_less_300.sum()
# Both Groupby and Sum is matching for Others category

4954

In [34]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

       rest_type                        cuisines  Cost 2 person    type  
0  Casual Dining  North Indian, Mughlai, Chinese          800.0  Buffet  
1  Casual Dining     Chinese, North Indian, Thai          800.0  Buffet  
2         Others          Cafe, Mexican, Italian          800.0  Buffet  
3    Quick Bites      South Indian, North Indian          300.0  Buffet  
4  Casual Dining        North Indian, Rajasthani          600.0  Buffet

# Name, Online Order, Book Table, Type and Votes columns check for cleaning

In [35]:
df.name.describe() # Same restaurants have multiple branches, as 8723 restaurants are present 51042 times

count               51042
unique               8723
top       Cafe Coffee Day
freq                   96
Name: name, dtype: object

In [36]:
df.name.unique() # No cleaning required

array(['Jalsa', 'Spice Elephant', 'San Churro Cafe', ..., 'Nawabs Empire',
       'SeeYa Restaurant',
       'Plunge - Sheraton Grand Bengaluru Whitefield Hotel &...'],
      dtype=object)

In [37]:
df.book_table.describe() # 2 unique value = Yes and No. No cleaning needed.

count     51042
unique        2
top          No
freq      44626
Name: book_table, dtype: object

In [38]:
df.online_order.describe() # 2 unique value = Yes and No. No cleaning needed.

count     51042
unique        2
top         Yes
freq      30228
Name: online_order, dtype: object

In [39]:
df.type.describe()

count        51042
unique           7
top       Delivery
freq         25579
Name: type, dtype: object

In [40]:
df.type.value_counts() #Good Distribution of data. Hence, no further action needed.

Delivery              25579
Dine-out              17562
Desserts               3559
Cafes                  1703
Drinks & nightlife     1084
Buffet                  869
Pubs and bars           686
Name: type, dtype: int64

In [41]:
df.votes.describe() #No cleaning needed

count    51042.000000
mean       284.507700
std        805.927478
min          0.000000
25%          7.000000
50%         41.000000
75%        199.000000
max      16832.000000
Name: votes, dtype: float64

In [42]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

       rest_type                        cuisines  Cost 2 person    type  
0  Casual Dining  North Indian, Mughlai, Chinese          800.0  Buffet  
1  Casual Dining     Chinese, North Indian, Thai          800.0  Buffet  
2         Others          Cafe, Mexican, Italian          800.0  Buffet  
3    Quick Bites      South Indian, North Indian          300.0  Buffet  
4  Casual Dining        North Indian, Rajasthani          600.0  Buffet

# Clean Cuisines column

In [43]:
df.cuisines.describe()

count            51042
unique            2704
top       North Indian
freq              2852
Name: cuisines, dtype: object

In [44]:
rest_cuis_group= df.groupby('cuisines').size()
rest_cuis_group

cuisines
African, Burger                                                      11
African, Burger, Desserts, Beverages, Fast Food                       2
American                                                             14
American, Asian, Continental, North Indian, South Indian, Chinese    12
American, Asian, European, North Indian                               8
                                                                     ..
Turkish, Fast Food, Biryani, Chinese                                  1
Turkish, Rolls                                                       17
Vietnamese                                                           23
Vietnamese, Salad                                                     2
Vietnamese, Thai, Burmese, Japanese                                   4
Length: 2704, dtype: int64

In [45]:
# Clubbing cuisines with less than 100 restaurants as Others, as 2635 of 2704 (around 97%) cuisines have more than 50% occurance.
print("Cuisines: ",rest_cuis_group[rest_cuis_group < 100].count())
print("Frequency: ",rest_cuis_group[rest_cuis_group < 100].sum())

Cuisines:  2635
Frequency:  26159


In [46]:
rest_less_100 = rest_cuis_group[rest_cuis_group < 100]
rest_less_100

cuisines
African, Burger                                                      11
African, Burger, Desserts, Beverages, Fast Food                       2
American                                                             14
American, Asian, Continental, North Indian, South Indian, Chinese    12
American, Asian, European, North Indian                               8
                                                                     ..
Turkish, Fast Food, Biryani, Chinese                                  1
Turkish, Rolls                                                       17
Vietnamese                                                           23
Vietnamese, Salad                                                     2
Vietnamese, Thai, Burmese, Japanese                                   4
Length: 2635, dtype: int64

In [47]:
def cuis_category(i):
    if i in rest_less_100.index:
        return "Others"
    else:
        return i
    
df.cuisines = df.cuisines.apply(cuis_category)   

df.groupby('cuisines').size().sort_values()

cuisines
Italian, Pizza                           102
North Indian, Chinese, Seafood           102
South Indian, Fast Food                  104
North Indian, Mughlai, Chinese           104
South Indian, Chinese, North Indian      105
                                       ...  
Biryani                                  903
South Indian                            1820
North Indian, Chinese                   2351
North Indian                            2852
Others                                 26159
Length: 70, dtype: int64

In [48]:
df.head()

name online_order book_table  rate  votes      location  \
0                  Jalsa          Yes        Yes   4.1    775  Banashankari   
1         Spice Elephant          Yes         No   4.1    787  Banashankari   
2        San Churro Cafe          Yes         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No   3.7     88  Banashankari   
4          Grand Village           No         No   3.8    166  Basavanagudi   

       rest_type                        cuisines  Cost 2 person    type  
0  Casual Dining  North Indian, Mughlai, Chinese          800.0  Buffet  
1  Casual Dining                          Others          800.0  Buffet  
2         Others                          Others          800.0  Buffet  
3    Quick Bites      South Indian, North Indian          300.0  Buffet  
4  Casual Dining                          Others          600.0  Buffet

# Saving cleaned dataset

In [49]:
df.to_excel('Zomato_clean_data.xlsx')